sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -58,77 +58,126 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val array = ctx.freshName("array")
-
     val et = dataType.elementType
     val evals = children.map(e => e.genCode(ctx))
     val isPrimitiveArray = ctx.isPrimitiveType(et)
-    val (preprocess, arrayData) =
-      GenArrayData.getCodeArrayData(ctx, et, children.size, isPrimitiveArray, array)
-
-    val assigns = if (isPrimitiveArray) {
-      val primitiveTypeName = ctx.primitiveTypeName(et)
-      evals.zipWithIndex.map { case (eval, i) =>
-        eval.code + s"""
-         if (${eval.isNull}) {
-           $arrayData.setNullAt($i);
-         } else {
-           $arrayData.set$primitiveTypeName($i, ${eval.value});
-         }
-       """
-      }
-    } else {
-      evals.zipWithIndex.map { case (eval, i) =>
-        eval.code + s"""
-         if (${eval.isNull}) {
-           $array[$i] = null;
-         } else {
-           $array[$i] = ${eval.value};
-         }
-       """
-      }
-    }
-    ev.copy(code =
-      preprocess +
-      ctx.splitExpressions(ctx.INPUT_ROW, assigns) +
-      s"\nfinal ArrayData ${ev.value} = $arrayData;\n",
+    val (preprocess, postprocess, arrayData, array) =
+      GenArrayData.genCodeToCreateArrayData(ctx, et, children.size, isPrimitiveArray)
+    val assigns = GenArrayData.genCodeToAssignArrayElements(
+      ctx, evals, et, isPrimitiveArray, arrayData, array, true)
+    /*
+      TODO: When we generate simpler code, we have to solve the following exception
+        https://github.com/apache/spark/pull/13909/files#r93813725
+      ev.copy(
+        code = preprocess + ctx.splitExpressions(ctx.INPUT_ROW, assigns) + postprocess
+        value = arrayData,
```

Comment:
Yes, it can avoid the exception. However, it may introduce a new performance issue. At lines 82 and 86, two castings happens per loop iteration. I would like to avoid these frequent castings.

```
/* 070 */     if (((ArrayData) arrayData1) instanceof UnsafeArrayData) {
/* 071 */
/* 072 */       final int sizeInBytes1 = ((UnsafeArrayData) ((ArrayData) arrayData1)).getSizeInBytes();
/* 073 */       // grow the global buffer before writing data.
/* 074 */       holder.grow(sizeInBytes1);
/* 075 */       ((UnsafeArrayData) ((ArrayData) arrayData1)).writeToMemory(holder.buffer, holder.cursor);
/* 076 */       holder.cursor += sizeInBytes1;
/* 077 */
/* 078 */     } else {
/* 079 */       final int numElements1 = ((ArrayData) arrayData1).numElements();
/* 080 */       arrayWriter1.initialize(holder, numElements1, 8);
/* 081 */
/* 082 */       for (int index1 = 0; index1 < numElements1; index1++) {
/* 083 */         if (((ArrayData) arrayData1).isNullAt(index1)) {
/* 084 */           arrayWriter1.setNullLong(index1);
/* 085 */         } else {
/* 086 */           final long element1 = ((ArrayData) arrayData1).getLong(index1);
/* 087 */           arrayWriter1.write(index1, element1);
/* 088 */         }
/* 089 */       }
/* 090 */     }
```